File: src/backend/SB.PortalSolicitudes/SB.PortalSolicitudes.Domain/Entities/Solicitud.cs
using SB.PortalSolicitudes.Domain.Common;

namespace SB.PortalSolicitudes.Domain.Entities;

/// <summary>
/// Entidad central del dominio: la solicitud interna registrada por un empleado.
/// No expone un campo de comentario de resolucion: ese comentario es el de la transicion
/// hacia <c>RESUELTA</c>, guardado en <see cref="HistorialEstado"/> (ver ADR-0001).
/// El cambio de estado y su registro en el historial los realiza la capa de aplicacion,
/// validando antes contra <see cref="TransicionPermitida"/> (ver ADR-0005).
/// </summary>
public class Solicitud : EntidadBase
{
    public const int MAX_LONGITUD_CODIGO = 20;
    public const int MAX_LONGITUD_TITULO = 150;
    public const int MAX_LONGITUD_DESCRIPCION = 2000;

    /// <summary>Codigo legible y unico, con el formato <c>SOL-2026-0001</c>.</summary>
    public required string Codigo { get; set; }

    public required string Titulo { get; set; }

    public required string Descripcion { get; set; }

    public DateTime FechaCreacion { get; set; }

    /// <summary>Fecha limite de atencion. Se captura manualmente y puede quedar sin definir.</summary>
    public DateTime? FechaCompromiso { get; set; }

    public int PrioridadId { get; set; }

    public Prioridad? Prioridad { get; set; }

    public int EstadoId { get; set; }

    public EstadoSolicitud? Estado { get; set; }

    public int AreaId { get; set; }

    public Area? Area { get; set; }

    public int TipoSolicitudId { get; set; }

    public TipoSolicitud? TipoSolicitud { get; set; }

    public int UsuarioSolicitanteId { get; set; }

    public Usuario? UsuarioSolicitante { get; set; }

    /// <summary>Responsable tecnico. Sin definir mientras la solicitud no se asigne.</summary>
    public int? UsuarioAsignadoId { get; set; }

    public Usuario? UsuarioAsignado { get; set; }

    public ICollection<HistorialEstado> Historial { get; set; } = new List<HistorialEstado>();

    public ICollection<Comentario> Comentarios { get; set; } = new List<Comentario>();

    public ICollection<Adjunto> Adjuntos { get; set; } = new List<Adjunto>();

    public ICollection<Notificacion> Notificaciones { get; set; } = new List<Notificacion>();
}

File: src/frontend/portal-solicitudes/src/types/solicitud.ts
import type { Area } from './area';
import type { EstadoSolicitud } from './estadoSolicitud';
import type { Prioridad } from './prioridad';
import type { TipoSolicitud } from './tipoSolicitud';
import type { UsuarioResumen } from './usuario';

/**
 * Forma exacta de SolicitudResumenDto (ver ADR-0030): fila de listado y dashboard, siempre con
 * objetos anidados, nunca ids planos — el DTO real no los trae. `descripcion` no viaja aqui,
 * solo en SolicitudDetalle.
 */
export interface Solicitud {
  id: number;
  /** Codigo legible y unico, formato SOL-2026-0001. */
  codigo: string;
  titulo: string;
  estado: EstadoSolicitud;
  prioridad: Prioridad;
  area: Area;
  tipoSolicitud: TipoSolicitud;
  solicitante: UsuarioResumen;
  asignado: UsuarioResumen | null;
  fechaCreacion: string;
  fechaCompromiso: string | null;
  /** Calculado en el servidor a partir de fechaCompromiso y estado — nunca recalcular en el cliente. */
  estaVencida: boolean;
}
